### PR TITLE
AzureCLI[V2/V3]: Fix special char corruption with dynamic PowerShell module

### PR DESCRIPTION
**Pull request by v-nairin**
## Summary

Replace the PowerShell `function az` interception with a dynamic module exporting an az alias to a physical az.ps1 shim. 

This:

- Bypasses `az.cmd` to preserve passwords containing CMD metacharacters such as `^`, `%`, `!`, `&`, `<`, and `>`.
- Makes `(Get-Command az).Source`, `.Path`, and `.Definition` identify the `az.ps1` shim.
- Invokes Azure CLI through `python.exe -IBm azure.cli`.
- Preserves native `$?` and `$LASTEXITCODE` behavior.
- Preserves caller-scoped `$ErrorActionPreference`, `$PSNativeCommandUseErrorActionPreference`, and `$PSNativeCommandArgumentPassing`.
- Saves and restores `AZ_INSTALLER`, including absent and empty states.
- Supports Windows PowerShell 5.1, PowerShell 7, and Constrained Language Mode.
- Restores an existing customer alias when the module is removed in Full Language mode.
- Cleans up partial imports and falls back to stock `az.cmd` when runtime module injection fails.
- Emits a visible warning when runtime fallback occurs.
- Uses best-effort shim cleanup so cleanup failures do not fail an otherwise successful task.
- Forwards piped/multiline stdin through $input to Azure CLI.

## Changes

Applied to AzureCLIV2 and AzureCLIV3:

- `Utility.ts`
  - Adds `getPowerShellScriptPathWithAzModule()`.
  - Creates a temporary `az.ps1` shim and dynamic module preamble.
  - Exports `az` as an alias to the shim.
  - Removes the old inline function from the legacy wrapper builder.
  - Adds partial-setup cleanup, runtime fallback, and telemetry.
  - Conditionally forwards pipeline input through $input while preserving no-input invocation behavior.g.

- `ScriptType.ts`
  - Selects file invocation only on Windows when `AzureCliUseFileInvocation` is enabled.
  - Falls back to the legacy `-Command` invocation if setup fails.
  - Tracks and removes wrapper and shim artifacts.

- Tests
  - Cover FF on/off, missing `az`, missing `python.exe`, shim-write failure, setup fallback, and runtime partial-import fallback.
  - Execute production-generated wrappers under Windows PowerShell 5.1 and PowerShell 7.
  - Cover Constrained Language Mode, command collisions, module unload, previous alias restoration, nested caller preferences, argument fidelity, `$?`, `$LASTEXITCODE`, and `AZ_INSTALLER` restoration.

The sanitizer test-isolation changes were moved to and merged separately in #22496.

## Versions

- AzureCLIV2: `2.280.0` → `2.280.2`
- AzureCLIV3: `3.280.0` → `3.280.2`

## Feature Flag

All new behavior is gated by `AzureCliUseFileInvocation`.

When the flag is disabled, PowerShell continues through the legacy `-Command` invocation. The original FF-on function injection has been removed from that legacy wrapper builder.

If file-invocation setup fails, the task falls back to the legacy invocation. If runtime module injection fails, the wrapper removes any partial module import, emits a warning, and continues with stock `az.cmd`.

## Testing

- AzureCLIV2 and AzureCLIV3 L0 suites pass locally.
- Node20 and Node24 generated variants validated.
- Windows PowerShell 5.1 and PowerShell 7 process tests pass.
- Private pipeline validation covers service-principal login, WIF, SQL passwords containing special characters, direct Python login, native status propagation, command collisions, CLM, fallback, and shim cleanup.

## Rollback

Disable the `AzureCliUseFileInvocation` feature flag.

### CODE REVIEW

File: Tasks/AzureCLIV2/src/Utility.ts [hunk 1]
```diff
@@ -154,4 +132,213 @@ export class Utility {
...
+
+            Utility.emitAzTelemetry('AzShimCreated', { status: 'created' });
+
+            // Module is named after the shim path so (Get-Command az).Source returns it.
+            // Do NOT rename to a human-friendly name — it breaks .Source resolution.
+            const lines = [
+                `$azureCliTaskShimPath = '${escapedShimPath}'`,
+                `$azureCliTaskPythonPath = '${escapedPythonPath}'`,
+                `try {`,
+                `    $azureCliTaskModule = New-Module -Name $azureCliTaskShimPath -ArgumentList $azureCliTaskPythonPath -ScriptBlock {`,
```
Comment (nemanjarogic) [inline Tasks/AzureCLIV2/src/Utility.ts:277]: I think the exported `az` function no longer observes caller-scoped PowerShell preferences.

Module functions execute in the module’s own session-state hierarchy, so a customer script that sets `$PSNativeCommandUseErrorActionPreference = $true` and `$ErrorActionPreference = 'Stop'` does not affect the native Python invocation on line 285.

I reproduced this with the real Azure CLI Python entry point: the previous inline function threw `NativeCommandExitException` for an invalid `az` command and skipped subsequent statements, whereas this module function continued.

A later successful native command then changed `$LASTEXITCODE` to 0, allowing the script/task to finish successfully despite the failed Azure operation. `$PSNativeCommandArgumentPassing` is isolated in the same way and changes empty/quoted argument serialization.

Please preserve the caller’s effective native-command preferences when invoking Python and add a process-level PowerShell 7 regression test where a failed `az` command must stop before a subsequent successful native command.

File: Tasks/AzureCLIV3/src/Utility.ts [hunk 2]
```diff
@@ -154,4 +132,213 @@ export class Utility {
...
+
+            Utility.emitAzTelemetry('AzShimCreated', { status: 'created' });
+
+            // Module is named after the shim path so (Get-Command az).Source returns it.
+            // Do NOT rename to a human-friendly name — it breaks .Source resolution.
+            const lines = [
+                `$azureCliTaskShimPath = '${escapedShimPath}'`,
+                `$azureCliTaskPythonPath = '${escapedPythonPath}'`,
+                `try {`,
+                `    $azureCliTaskModule = New-Module -Name $azureCliTaskShimPath -ArgumentList $azureCliTaskPythonPath -ScriptBlock {`,
```
Comment (nemanjarogic) [inline Tasks/AzureCLIV3/src/Utility.ts:277]: Please see [this comment](https://github.com/microsoft/azure-pipelines-tasks/pull/22452/changes#r3902341206).

File: Tasks/AzureCLIV2/Tests/L0.ts [hunk 3]
```diff
@@ -314,8 +314,8 @@ describe('AzureCLIV2 Suite', function () {
         assert(!tr.succeeded, 'task should have failed due to stderr with failOnStandardError=true');
     });
 
-    it('Az function alias injection: injects function when FF on + az found + python.exe exists', async () => {
-        let tp = path.join(__dirname, 'L0AzFunctionAliasInjection.js');
+    it('Az module injection: injects dynamic module when FF on + az found + python.exe exists', async () => {
```
Comment (nemanjarogic) [inline Tasks/AzureCLIV2/Tests/L0.ts:317]: These assertions validate the generated text but never execute the generated PowerShell. As a result, the central runtime contract - module import, `(Get-Command az).Source/.Path`, caller preference visibility, real argument forwarding, environment restoration, and native error behavior is currently uncovered.

Please add a process-level test that launches the generated wrapper under `pwsh` and Windows PowerShell. In particular, `include $PSNativeCommandUseErrorActionPreference = $true`, a failing native `az`, and a following command that must not execute. That scenario exposes a current regression in the module implementation.

File: Tasks/AzureCLIV3/Tests/L0.ts [hunk 4]
```diff
@@ -596,8 +596,8 @@ describe('AzureCLIV3 Suite', function () {
         assert(tr.stdout.indexOf('TELEMETRY: AzureCLIV3/DirectPythonLogin') >= 0, 'should emit DirectPythonLogin telemetry');
     });
 
-    it('Az function alias injection: injects function when FF on + az found + python.exe exists', async () => {
-        let tp = path.join(__dirname, 'L0AzFunctionAliasInjection.js');
+    it('Az module injection: injects dynamic module when FF on + az found + python.exe exists', async () => {
```
Comment (nemanjarogic) [inline Tasks/AzureCLIV3/Tests/L0.ts:599]: See [this comment](https://github.com/microsoft/azure-pipelines-tasks/pull/22452/changes#r3902361068) for V2.

File: Tasks/AzureCLIV2/Tests/L0AzModuleShimWriteFailure.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+process.env['AGENT_TEMPDIRECTORY'] = os.tmpdir();
+
+tmr.registerMock('os', {
+    platform: () => 'win32',
+    tmpdir: () => os.tmpdir(),
+    EOL: os.EOL
+});
+
+const realFs = require('fs');
+let rmRFCalled: string[] = [];
```
Comment (nemanjarogic) [inline Tasks/AzureCLIV2/Tests/L0AzModuleShimWriteFailure.ts]: This test is named “shim write failure,” but `writeFileSync` is a pass-through and `mkdtempSync` throws before any shim directory or file exists. It therefore never exercises the shim-write failure or partial-setup cleanup path, and `rmRFCalled` is not used in an assertion.

Please let `mkdtempSync` create/return a real temporary directory, throw specifically when `writeFileSync` targets `az.ps1`, and assert that `rmRF` receives that exact directory and that no partial shim remains.

Comment (nemanjarogic) [reply Tasks/AzureCLIV2/Tests/L0AzModuleShimWriteFailure.ts]: The same comment goes for AzureCLIV3

File: Tasks/AzureCLIV2/src/Utility.ts (hunk 1 above)
Comment (v-nairin) [reply Tasks/AzureCLIV2/src/Utility.ts:277]: Added the change. The az module now forwards the caller-scoped PowerShell preferences to the native Python invocation. I also added a process-level PowerShell 7 test covering empty/quoted arguments and verifying that a failed az raises NativeCommandExitException and prevents the subsequent successful az command from running.

File: Tasks/AzureCLIV3/src/Utility.ts (hunk 2 above)
Comment (v-nairin) [reply Tasks/AzureCLIV3/src/Utility.ts:277]: Fixed for AzureCLIV3 as well in 4a543a7e70, with the same caller-preference forwarding and regression coverage.

File: Tasks/AzureCLIV2/Tests/L0.ts (hunk 3 above)
Comment (v-nairin) [reply Tasks/AzureCLIV2/Tests/L0.ts:317]: Updated . The test executes the production-generated wrapper under PowerShell 7 and Windows PowerShell. It covers command metadata, argument forwarding, environment restoration, and native failure behavior. The V2 L0 suite passes.

File: Tasks/AzureCLIV3/Tests/L0.ts (hunk 4 above)
Comment (v-nairin) [reply Tasks/AzureCLIV3/Tests/L0.ts:599]: Updated for AzureCLIV3 in 4a543a7e70 with equivalent PowerShell 7 and Windows PowerShell coverage. The V3 L0 suite passes.

File: Tasks/AzureCLIV2/Tests/L0AzModuleShimWriteFailure.ts (hunk 5 above)
Comment (v-nairin) [reply Tasks/AzureCLIV2/Tests/L0AzModuleShimWriteFailure.ts]: Fixed for both V2 and V3 in . The test creates a real temporary directory, fails specifically while writing az.ps1, verifies rmRF receives that exact directory, and confirms no partial shim remains.

**Comment by wawanawna** [comment]
Overall this looks solid - the logic, tests, feature-flag gating and the fallback all hold up, and I don't see functional blockers. The one thing worth tightening before merge is **duplication**, since the new az-shim code is non-trivial and easy to let drift:

**1. AzureCLIV2 and V3 are identical copies.** `src/Utility.ts` and `src/ScriptType.ts` are byte-for-byte identical across V2 and V3 except the `emitTelemetry('AzureCLIV2' / 'AzureCLIV3')` literal — ~300 net-new lines duplicated. A future fix to the shim/preamble has to be made in both (and re-generated under `_generated/`). This task already consumes a shared module (`../Common/ArgsSanitizer` in `make.json`), so `createAzModulePreamble` / `getPowerShellScriptPathWithAzModule` / `deleteDirectory` / `emitAzTelemetry` could move to `Tasks/Common/` with the telemetry area passed in. (The pre-existing `getSriptType` vs `getScriptType` typo between V2 and V3 shows the copies already drift.)

**2. `getPowerShellScriptPathWithAzModule` repeats `getPowerShellScriptPath`.** In the same file it re-implements the error-action validation, `assertAgent`, temp-dir setup, the `$ErrorActionPreference`/`$ErrorView` header, inline-script file creation, the `$LASTEXITCODE` trailer and the BOM write. Could collapse into one builder that takes optional preamble lines and returns the optional `azShimDirectory`.

**3. `WindowsPowerShell` and `PowerShellCore` are near-identical.** `getTool` / `getToolWithFileInvocation` / `cleanUp` differ only by `powershell` vs `pwsh` and the `ps`/`pscore` tag; the feature-gate + try/catch/fallback is copy-pasted. That shared body could move into the abstract `ScriptType` base.

File: Tasks/AzureCLIV2/src/Utility.ts [hunk 6]
```diff
@@ -154,4 +132,220 @@ export class Utility {
...
+                `        param([string] $pythonPath)`,
+                `        $script:pythonPath = $pythonPath`,
+                `        function az {`,
+                `            $callerFrame = (Get-PSCallStack)[1].GetFrameVariables()`,
```
Comment (nemanjarogic) [inline Tasks/AzureCLIV2/src/Utility.ts]: Are you sure this will work? Copilot is telling me that this makes `az` unusable in Constrained Language Mode since `CallStackFrame.GetFrameVariables()` is not permitted in ConstrainedLanguage.

Comment (nemanjarogic) [reply Tasks/AzureCLIV2/src/Utility.ts]: The same goes for V3

Comment (nemanjarogic) [reply Tasks/AzureCLIV2/src/Utility.ts]: Also might be related: Please module-qualify the built-in `Get-PSCallStack` command. Copilot verified with the exact generated wrapper that a customer-defined `Get-PSCallStack` function is resolved here instead, causing every `az` call to throw under both Windows PowerShell and PowerShell 7. `Microsoft.PowerShell.Utility\Get-PSCallStack` avoided the shadowing in both hosts.

Comment (starkmsu) [inline Tasks/AzureCLIV2/src/Utility.ts]: This still loses dynamically inherited caller preferences when `az` is invoked through a nested helper. `GetFrameVariables()` on frame `[1]` contains only the immediate caller's variables, so preferences set by an outer function are missed. I reproduced this under PowerShell 7: an outer function set `$ErrorActionPreference = 'Stop'` and `$PSNativeCommandUseErrorActionPreference = $true`, an inner function called `az`, both keys were absent from the inspected frame, and execution continued after exit code 7.

This is a recurrence of the scenario from https://github.com/microsoft/azure-pipelines-tasks/pull/22452#discussion_r3902341206 after the forwarding fix. Please walk outward through the call stack and use the nearest frame defining each preference, with an outer/inner caller regression test. The mirrored V3 implementation has the same issue.

File: Tasks/AzureCLIV2/src/Utility.ts [hunk 7]
```diff
@@ -154,4 +132,220 @@ export class Utility {
...
+                `            }`,
+                `            finally {`,
+                `                if ($null -eq $previousInstaller) {`,
+                `                    Remove-Item Env:\\AZ_INSTALLER -ErrorAction SilentlyContinue`,
+                `                }`,
+                `                else {`,
+                `                    $env:AZ_INSTALLER = $previousInstaller`,
+                `                }`,
+                `            }`,
+                `            $global:LASTEXITCODE = $azExitCode`,
```
Comment (starkmsu) [inline Tasks/AzureCLIV2/src/Utility.ts]: A nonzero `az` invocation now returns with `$LASTEXITCODE` set correctly but `$?` set to `True`. After Python exits, the function performs successful assignments and environment-restoration commands, ending with this assignment; I reproduced `SUCCESS=True EXIT=7` under both Windows PowerShell 5.1 and PowerShell 7.

That breaks scripts using `if ($?)` or `if (-not $?)`, which can run success-only logic after a failed Azure CLI operation. Please preserve native-command success semantics as well as `$LASTEXITCODE`—for example, make invocation of the existing `az.ps1` shim the function's final operation—and add assertions for both values after failure. The mirrored V3 implementation has the same issue.

File: Tasks/AzureCLIV2/src/Utility.ts (hunk 6 above)
Comment (v-nairin) [reply Tasks/AzureCLIV2/src/Utility.ts]: Addressed by replacing the exported function with an exported alias targeting the physical az.ps1 shim. The implementation no longer uses Get-PSCallStack or GetFrameVariables(). Process tests now cover Constrained Language Mode under Windows PowerShell 5.1 and PowerShell 7 for both V2 and V3.
Changed to an alias because a PowerShell function boundary could not reliably preserve native-command semantics, especially $?, inherited caller preferences, and CLM compatibility. An alias to the physical shim behaves like direct script invocation while still providing the required (Get-Command az).Source path.

Comment (v-nairin) [reply Tasks/AzureCLIV2/src/Utility.ts]: Addressed by replacing the function boundary with an alias to az.ps1. PowerShell now naturally carries dynamically scoped preferences through nested callers, without inspecting stack frames. The regression test sets preferences in an outer function and calls az from an inner function under PowerShell 7.

File: Tasks/AzureCLIV2/src/Utility.ts (hunk 7 above)
Comment (v-nairin) [reply Tasks/AzureCLIV2/src/Utility.ts]: Addressed by exporting an alias directly to the physical az.ps1 shim. This removes the function boundary that normalized $?. Process tests under Windows PowerShell 5.1 and PowerShell 7 assert $? = False and $LASTEXITCODE = 7 after a failing az call, including calls with piped stdin.

**Comment by v-nairin** [comment]
https://github.com/microsoft/azure-pipelines-tasks/pull/22452#issuecomment-5581492902

Thanks Vova. I agree there is duplication across V2/V3 and the PowerShell implementations. I’m deferring that refactor because it would broaden this bug fix and touch the legacy feature-flag-off path in master.

**Review by nemanjarogic — APPROVED** [review]

**Review by starkmsu — APPROVED** [review]
